cmd/opampsupervisor/supervisor/packages.go

```diff
@@ -0,0 +1,417 @@
+		return nil, nil, fmt.Errorf("signature must be formatted as a space separated cert and signature")
+	}
+
+	return splitSignature[0], splitSignature[1], nil
+}
+
+// verifyPackageSignature verifies that the b64Signature is a valid signature for packageBytes.
+// b64Cert is used to validate the identity of the signature against the identities in the
+// provided checkOpts.
+func verifyPackageSignature(ctx context.Context, checkOpts *cosign.CheckOpts, packageBytes, b64Cert, b64Signature []byte) error {
+	decodedCert, err := base64.StdEncoding.AppendDecode(nil, b64Cert)
+	if err != nil {
+		return fmt.Errorf("b64 decode cert: %w", err)
+	}
+
+	ociSig, err := static.NewSignature(packageBytes, string(b64Signature), static.WithCertChain(decodedCert, nil))
+	if err != nil {
+		return fmt.Errorf("create signature: %w", err)
+	}
+
+	_, err = cosign.VerifyBlobSignature(ctx, ociSig, checkOpts)
+	if err != nil {
+		return fmt.Errorf("verify blob: %w", err)
+	}
+
+	return nil
+}
+
+func createCosignCheckOpts(signatureOpts config.AgentSignature) (*cosign.CheckOpts, error) {
+	rootCerts, err := fulcio.GetRoots()
+	if err != nil {
+		return nil, fmt.Errorf("fetch root certs: %w", err)
+	}
+
+	intermediateCerts, err := fulcio.GetIntermediates()
+	if err != nil {
+		return nil, fmt.Errorf("fetch intermediate certs: %w", err)
+	}
+
+	rekorClient, err := client.GetRekorClient("https://rekor.sigstore.dev")
+	if err != nil {
+		return nil, fmt.Errorf("create rekot client: %w", err)
+	}
+
+	rekorKeys, err := cosign.GetRekorPubs(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("get rekor public keys: %w", err)
+	}
+
+	ctLogPubKeys, err := cosign.GetCTLogPubs(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("get CT log public keys: %w", err)
+	}
+
+	identities := make([]cosign.Identity, 0, len(signatureOpts.Identities))
+	for _, ident := range signatureOpts.Identities {
+		identities = append(identities, cosign.Identity{
+			Issuer:        ident.Issuer,
+			IssuerRegExp:  ident.IssuerRegExp,
+			Subject:       ident.Subject,
+			SubjectRegExp: ident.SubjectRegExp,
+		})
+	}
+
+	return &cosign.CheckOpts{
+		RootCerts:                    rootCerts,
+		IntermediateCerts:            intermediateCerts,
+		CertGithubWorkflowRepository: signatureOpts.CertGithubWorkflowRepository,
+		Identities:                   identities,
+		RekorClient:                  rekorClient,
+		RekorPubKeys:                 rekorKeys,
+		CTLogPubKeys:                 ctLogPubKeys,
+	}, nil
+}
+
+func restoreBackup(backupPath, restorePath string) error {
+	backupFile, err := os.Open(backupPath)
+	if err != nil {
+		return fmt.Errorf("open backup file: %w", err)
+	}
+	defer backupFile.Close()
+
+	restoreFile, err := os.OpenFile(restorePath, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, 0600)
+	if err != nil {
+		return fmt.Errorf("open restore file: %w", err)
+	}
+	defer restoreFile.Close()
+
+	if _, err := io.Copy(restoreFile, backupFile); err != nil {
```

Comment:
Do we need to keep the backup file or we can rename instead of copying? Renaming likely is less susceptible to out-of-disk-space issues.